site/content/docs/reference/code-walkthrough.md

```diff
@@ -0,0 +1,216 @@
+conclusions, caused by one controller's caches lagging slightly behind or by new business logic introduced
+to an existing controller during a rolling upgrade.
+
+Leader election is done transparently in a centralized client middleware
+component, and will not be immediately obvious when looking at the controller source code. The middleware is in
+[internal/leaderelection/leaderelection.go](https://github.com/vmware-tanzu/pinniped/blob/main/internal/leaderelection/leaderelection.go).
+
+One of the Pods is always elected as leader, and only the Kubernetes API client calls made by that pod are allowed
+to perform writes. The non-leader Pods' controller are always running, but they will always result in failed writes,
+so they are unable to compete to make changes to Kubernetes resources. These failed write operations will appear in the logs
+as write errors due to not being the leader. While this might look like an error, this is normal for the controllers.
+This still allows the non-leader Pods' controllers to read state, update in-memory caches, etc.
+
+## Concierge API endpoints
+
+The Concierge hosts the following endpoints, which are automatically registered with the Kubernetes API server
+as aggregated API endpoints, which makes them appear to a client almost as if they were built into Kubernetes itself.
+
+- `TokenCredentialRequest` can receive a token, pass the token to an external authenticator to authenticate the user,
+  and then return a short-lived mTLS client certificate keypair which can be used to gain access to the Kuberetes API
+  as that user.
+  It is in [internal/registry/credentialrequest/rest.go](https://github.com/vmware-tanzu/pinniped/blob/main/internal/registry/credentialrequest/rest.go).
+
+- `WhoAmI` will return basic details about the currently authenticated user.
+  It is in [internal/registry/whoamirequest/rest.go](https://github.com/vmware-tanzu/pinniped/blob/main/internal/registry/whoamirequest/rest.go).
+
+The Concierge may also run an impersonation proxy service. This is not an aggregated API endpoint, so it needs to be
+exposed outside the cluster as a Service. When operating this mode, a client's kubeconfig causes the client to
+make all Kubernetes API requests to the impersonation proxy endpoint instead of the real API server. The impersonation
+proxy then authenticates the user and calls the real Kubernetes API on their behalf.
+Calls made to the real API server are made as a service account using impersonation to impersonate the identity of the end user.
+The code tries to reuse as much code from Kubernetes itself as possible, so it can behave as closely as possible
+to the real API server from the client's point of view. It can be found in
+[internal/concierge/impersonator/impersonator.go](https://github.com/vmware-tanzu/pinniped/blob/main/internal/concierge/impersonator/impersonator.go).
+
+## Supervisor API endpoints
+
+The Supervisor's endpoints are:
```

Comment:
Maybe it would be helpful to link to the OIDC spec here.

Reply:
Updated!